src/ai/defaultai.cc

```diff
@@ -3341,70 +3341,202 @@ void DefaultAI::diplomacy_actions(const Time& gametime) {
+	// first.
+	// If alone, we may accept requests to join and cancel leaving.
+	// If defeated, just clean up by rejecting everything.
 	if (me->team_number() != 0 && (me_alone || me_def)) {
-		game().send_player_diplomacy(mypn, Widelands::DiplomacyAction::kLeaveTeam, 0 /* ignored */);
-		verb_log_dbg_time(gametime, "AI Diplomacy: Player(%d) leaves team (%d) as %s.\n",
-		                  static_cast<unsigned int>(mypn), static_cast<unsigned int>(mytn),
-		                  me_alone ? "last one" : "defeated");
+		planned_action = Widelands::DiplomacyAction::kLeaveTeam;
+		have_plan = true;
+		plan_priority = me_alone ? 0 : std::numeric_limits<int32_t>::max();
+		if (g_verbose) {
+			planned_log_append_text = me_alone ? " as last one" : " as defeated";
+			/* verb_ */ log_dbg_time(gametime,
+			                         "AI Diplomacy: Player(%d) plans to leave team with priority %d",
+			                         static_cast<unsigned int>(mypn), plan_priority);
+		}
 	}
 
 	const int32_t my_team_score =
 	   me_alone ? 0 : player_statistics.get_team_average_score(mytn, mypn);
 	const std::string myts_s = me_alone ? "" : format(" with team score %d", my_team_score);
 
-	bool request_accepted = false;
+	// Check desirability of current team
+	if (!have_plan && !me_alone && my_team_score < 0 &&
+	    (my_team_score < -10 || RNG::static_rand(8) == 0)) {
+		planned_action = Widelands::DiplomacyAction::kLeaveTeam;
+		have_plan = true;
+		plan_priority = -my_team_score;
+		if (g_verbose) {
+			planned_log_append_text = myts_s;
+			/* verb_ */ log_dbg_time(gametime,
+			                         "AI Diplomacy: Player(%d) plans to leave team with priority %d",
+			                         static_cast<unsigned int>(mypn), plan_priority);
+		}
+	}
+
+	// Check for undesirable teammates
+	if (!have_plan && player_statistics.get_worst_ally_score() < -15) {
+		const int32_t team_vs_worst = my_team_score + player_statistics.get_worst_ally_score();
+		if ((team_vs_worst > 0 && player_statistics.members_in_team(mytn) > 3) ||
+		    (team_vs_worst > -10 &&
+		     RNG::static_rand(player_statistics.members_in_team(mytn) * 2 - 2) > 0)) {
+			verb_log_dbg_time(gametime,
+			                  "AI Diplomacy: Player(%d) tolerates player (%d) with diploscore "
+			                  "%d in team (%d)%s\n",
+			                  static_cast<unsigned int>(mypn),
+			                  static_cast<unsigned int>(player_statistics.get_worst_ally()),
+			                  player_statistics.get_worst_ally_score(),
+			                  static_cast<unsigned int>(mytn), myts_s.c_str());
+		} else {
+			planned_action = Widelands::DiplomacyAction::kLeaveTeam;
+			have_plan = true;
+			plan_priority = std::max(0, -team_vs_worst);
+			if (g_verbose) {
+				planned_log_append_text =
+				   format("%s because of player (%d) with diploscore %d", myts_s,
+				          static_cast<unsigned int>(player_statistics.get_worst_ally()),
+				          player_statistics.get_worst_ally());
```

Comment:
Already corrected locally to `get_worst_ally_score_()`, but waiting for result of clang-tidy.